### PR TITLE
Make ArrowStyle docstrings numpydoc compatible

### PR DESCRIPTION
**Pull request by patniharshit**
refs #8342

### CODE REVIEW

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -3418,17 +3417,18 @@ def __init__(self):
...
+        """An arrow with a head at its begin point.
         """
 
         def __init__(self, head_length=.4, head_width=.2):
             """
-            *head_length*
-              length of the arrow head
+            Parameters
+            ----------
+            head_length : float, optional
```
Comment (NelleV) [inline lib/matplotlib/patches.py]: can you add the default value as well?
```
head_length : float, optional, default: 0.4
```

File: lib/matplotlib/patches.py [hunk 2]
```diff
@@ -3407,8 +3407,7 @@ def transmute(self, path, mutation_size, linewidth):
             return _path, _fillable
 
     class Curve(_Curve):
-        """
-        A simple curve without any arrow head.
+        """A simple curve without any arrow head.
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: Are all these single-line changes really necessary? Personally, I don't think they look very good like this.

File: lib/matplotlib/patches.py [hunk 3]
```diff
@@ -3418,17 +3417,18 @@ def __init__(self):
...
         """
 
         def __init__(self, head_length=.4, head_width=.2):
             """
-            *head_length*
-              length of the arrow head
+            Parameters
+            ----------
+            head_length : float, optional
+                        Length of the arrow head
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: 4-space indent, please.

File: lib/matplotlib/patches.py [hunk 4]
```diff
@@ -3520,18 +3524,19 @@ def __init__(self, head_length=.4, head_width=.2):
     _style_list["-|>"] = CurveFilledB
 
     class CurveFilledAB(_Curve):
-        """
-        An arrow with filled triangle heads both at the begin and the end
+        """An arrow with filled triangle heads both at the begin and the end
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: This really should be one line. Maybe change to "at both ends" like `BracketAB` does below.

File: lib/matplotlib/patches.py (hunk 1 above)
Comment (patniharshit) [inline lib/matplotlib/patches.py]: Added

File: lib/matplotlib/patches.py (hunk 3 above)
Comment (patniharshit) [inline lib/matplotlib/patches.py]: Sorry for that, I should not have mixed tabs and spaces.

File: lib/matplotlib/patches.py (hunk 2 above)
Comment (patniharshit) [inline lib/matplotlib/patches.py]: Yes they don't look good at all. But I think somewhere I saw that for single line docstring.

File: lib/matplotlib/patches.py [hunk 5]
```diff
@@ -3505,11 +3513,13 @@ class CurveFilledB(_Curve):
...
-            *head_length*
-              length of the arrow head
+            Parameters
+            ----------
+            head_length : float, optional, default : .4
+                        Length of the arrow head
 
-            *head_width*
-              width of the arrow head
+            head_width : float, optional, default : .2
```
Comment (phobson) [inline lib/matplotlib/patches.py]: For all of these, the numbers need a leading zero before the decimal (e.g., ".2" -> "0.2")

File: lib/matplotlib/patches.py [hunk 6]
```diff
@@ -3505,11 +3513,13 @@ class CurveFilledB(_Curve):
 
         def __init__(self, head_length=.4, head_width=.2):
             """
-            *head_length*
-              length of the arrow head
+            Parameters
+            ----------
+            head_length : float, optional, default : .4
```
Comment (phobson) [inline lib/matplotlib/patches.py]: cc @NelleV 

I don't we should have two colons in the same sentence. It looks weird to me (but it's not a deal breaker). I think this should read:

```
headhead_length : float, optional (default is 0.4)
```

The other question is if these parameters are actually optional. What I mean by that is if you said `head_length=None`, would this still work by falling back to the default value? If not, then the parameter isn't actually, optional, just named.

Comment (QuLogic) [inline lib/matplotlib/patches.py]: Optional does not imply that passing `None` does the default; it implies that not passing it uses the default.

File: lib/matplotlib/patches.py [hunk 7]
```diff
@@ -3424,11 +3424,13 @@ class CurveA(_Curve):
...
+                        Length of the arrow head
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: This should be indented only 4 more spaces from the previous line.

File: lib/matplotlib/patches.py (hunk 6 above)
Comment (phobson) [inline lib/matplotlib/patches.py]: @QuLogic 

You are correct. From the numpydoc spec:

> If it is not necessary to specify a keyword argument, use optional:
> 
> `x : int, optional`
> Optional keyword parameters have default values, which are displayed as part of the function signature. They can also be detailed in the description:
> 
```
Description of parameter `x` (the default is -1, which implies summation
over all axes).
If it is not necessary to specify a keyword argument, use optional:
```
> 
> `x : int, optional`
> Optional keyword parameters have default values, which are displayed as part of the function signature. They can also be detailed in the description:
>  
> `Description of parameter `x` (the default is -1, which implies summation over all axes).`

(though I wish the meaning of "optional" in this context was a little more precise.

File: lib/matplotlib/patches.py (hunk 7 above)
Comment (patniharshit) [inline lib/matplotlib/patches.py]: Here only or everywhere?

File: lib/matplotlib/patches.py (hunk 5 above)
Comment (patniharshit) [inline lib/matplotlib/patches.py]: Done.

File: lib/matplotlib/patches.py (hunk 6 above)
Comment (patniharshit) [inline lib/matplotlib/patches.py]: Should I change the way defaults are shown then?

File: lib/matplotlib/patches.py (hunk 7 above)
Comment (phobson) [inline lib/matplotlib/patches.py]: everywhere. Should like:

```
...
            Parameters
            ----------
            head_length : float, optional, default : .4
                Length of the arrow head
...
```

File: lib/matplotlib/patches.py [hunk 8]
```diff
@@ -3710,23 +3727,25 @@ def __init__(self,
...
-            *widthA*
-              width of the bracket
+            Parameters
+            ----------
+            widthA : float, optional, default : 1.0
+                Width of the bracket
 
-            *lengthA*
-              length of the bracket
+            lengthA : int
```
Comment (dstansby) [inline lib/matplotlib/patches.py]: I'm slightly confused, `__init__` just above this docstring doesn't take a `lengthA` or `lengthB` argument, is the docstring just wrong here at the moment?

File: lib/matplotlib/patches.py (hunk 6 above)
Comment (phobson) [inline lib/matplotlib/patches.py]: That would be my preference, but I'm not going to hold up this PR for that.

File: lib/matplotlib/patches.py (hunk 8 above)
Comment (phobson) [inline lib/matplotlib/patches.py]: I think you're right @dstansby .

@patniharshit can you update this docstring to more accurately reflect the call signature?

**Comment by patniharshit** [comment]
Happy to contribute :)

**Pull request merged**